### PR TITLE
MULE-10205: Create HTTP service. First iteration, most required builders missing.

### PR DESCRIPTION
**Pull request by afelisatti**

### CODE REVIEW

**Comment by afelisatti** [comment]
REVIEW ONLY, for now. Tests are still running and specific TODOs to come.

File: distributions/standalone/assembly-whitelist.txt [hunk 1]
```diff
@@ -183,6 +184,7 @@
 #
 /mule-standalone-${productVersion}/services
 /mule-standalone-${productVersion}/services/mule-service-scheduler-${productVersion}.zip
+/mule-standalone-${productVersion}/services/mule-http-service-impl-${productVersion}.zip
```
Comment (elrodro83) [inline distributions/standalone/assembly-whitelist.txt]: http-service or service-http? better just mule-service-http

File: distributions/standalone/assembly-whitelist.txt [hunk 2]
```diff
@@ -162,6 +162,7 @@
 /mule-standalone-${productVersion}/lib/mule/mule-module-tooling-support-${productVersion}.jar
 /mule-standalone-${productVersion}/lib/mule/mule-module-xml-${productVersion}.jar
 /mule-standalone-${productVersion}/lib/mule/mule-mvel2-2.1.9-MULE-010.jar
+/mule-standalone-${productVersion}/lib/mule/mule-service-http-${productVersion}.jar
```
Comment (elrodro83) [inline distributions/standalone/assembly-whitelist.txt]: mule-service-http-api?

File: modules/http-impl/src/main/java/org/mule/services/http/impl/provider/HttpServiceProvider.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+import org.mule.service.http.api.HttpService;
+import org.mule.services.http.impl.service.HttpServiceImplementation;
+
+import com.google.common.collect.Lists;
+
+import java.util.List;
+
+public class HttpServiceProvider implements ServiceProvider {
+
+  HttpServiceImplementation service = new HttpServiceImplementation();
```
Comment (elrodro83) [inline modules/http-impl/src/main/java/org/mule/services/http/impl/provider/HttpServiceProvider.java]: private

File: modules/http-impl/src/main/java/org/mule/services/http/impl/provider/HttpServiceProvider.java [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+  private ServiceDefinition serviceDefinition = new ServiceDefinition(HttpService.class, service);
+
+  @Override
+  public List<ServiceDefinition> providedServices() {
+    return Lists.newArrayList(serviceDefinition);
```
Comment (elrodro83) [inline modules/http-impl/src/main/java/org/mule/services/http/impl/provider/HttpServiceProvider.java]: singletonList?

Comment (elrodro83) [inline modules/http-impl/src/main/java/org/mule/services/http/impl/provider/HttpServiceProvider.java]: In the scheduler service, the ServiceDefinition is built in this method. We should confirm what's the proper way.

File: modules/http-impl/src/main/java/org/mule/services/http/impl/service/HttpServiceImplementation.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+import org.mule.service.http.api.HttpService;
+import org.mule.service.http.api.client.HttpClientFactory;
+import org.mule.service.http.api.server.HttpServerFactory;
+import org.mule.services.http.impl.service.client.GrizzlyHttpClient;
+import org.mule.services.http.impl.service.server.HttpListenerConnectionManager;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class HttpServiceImplementation implements HttpService, Startable, Stoppable {
```
Comment (elrodro83) [inline modules/http-impl/src/main/java/org/mule/services/http/impl/service/HttpServiceImplementation.java]: javadoc

File: modules/http-impl/src/main/java/org/mule/services/http/impl/service/client/GrizzlyHttpClient.java [hunk 6]
```diff
@@ -72,8 +69,6 @@
 
   private static final Logger logger = LoggerFactory.getLogger(GrizzlyHttpClient.class);
 
-  private UriParameters uriParameters;
```
Comment (elrodro83) [inline modules/http-impl/src/main/java/org/mule/services/http/impl/service/client/GrizzlyHttpClient.java]: is it ok to remove this?

File: modules/http-impl/src/main/java/org/mule/services/http/impl/service/server/HttpListenerConnectionManager.java [hunk 7]
```diff
@@ -60,18 +53,10 @@ public void initialise() throws InitialisationException {
...
-        muleContext.getRegistry().lookupObjects(TcpServerSocketProperties.class);
-    TcpServerSocketProperties tcpServerSocketProperties = new TcpServerSocketProperties();
-
-    if (tcpServerSocketPropertiesBeans.size() == 1) {
-      tcpServerSocketProperties = Iterables.getOnlyElement(tcpServerSocketPropertiesBeans);
-    } else if (tcpServerSocketPropertiesBeans.size() > 1) {
-      throw new InitialisationException(CoreMessages
-          .createStaticMessage("Only one global TCP server socket properties bean should be defined in the config"), this);
-    }
+    //TODO: Analyze how to allow users to configure this
```
Comment (elrodro83) [inline modules/http-impl/src/main/java/org/mule/services/http/impl/service/server/HttpListenerConnectionManager.java]: is there an issue for this?

File: modules/http-impl/src/main/java/org/mule/services/http/impl/service/server/HttpListenerConnectionManager.java [hunk 8]
```diff
@@ -60,18 +53,10 @@ public void initialise() throws InitialisationException {
...
+    TcpServerSocketProperties tcpServerSocketProperties = new DefaultTcpServerSocketProperties();
 
-    String threadNamePrefix = ThreadNameHelper.getPrefix(muleContext) + LISTENER_THREAD_NAME_PREFIX;
+    String threadNamePrefix = LISTENER_THREAD_NAME_PREFIX;
```
Comment (elrodro83) [inline modules/http-impl/src/main/java/org/mule/services/http/impl/service/server/HttpListenerConnectionManager.java]: add TODO MULE-11018

File: modules/http-impl/src/main/java/org/mule/services/http/impl/service/server/grizzly/GrizzlyServerManager.java [hunk 9]
```diff
@@ -104,10 +98,7 @@ public GrizzlyServerManager(String threadNamePrefix, HttpListenerRegistry httpLi
     // Set filterchain as a Transport Processor
     transport.setProcessor(serverFilterChainBuilder.build());
 
-    idleTimeoutExecutorService =
```
Comment (elrodro83) [inline modules/http-impl/src/main/java/org/mule/services/http/impl/service/server/grizzly/GrizzlyServerManager.java]: is this ok to remove?

File: modules/http-impl/src/main/java/org/mule/services/http/impl/service/server/grizzly/GrizzlyServerManager.java [hunk 10]
```diff
@@ -255,11 +249,43 @@ private int retrieveMaximumHeaderSectionSize() {
...
-                                                                                    MAXIMUM_HEADER_SECTION_SIZE_PROPERTY_KEY)),
+      throw new MuleRuntimeException(createStaticMessage(String.format("Invalid value %s for %s configuration",
+                                                                       getProperty(MAXIMUM_HEADER_SECTION_SIZE_PROPERTY_KEY),
+                                                                       MAXIMUM_HEADER_SECTION_SIZE_PROPERTY_KEY)),
                                      e);
     }
   }
 
+  private int convertToSeconds(int milliseconds) {
+    return (int) Math.ceil((double) milliseconds / 1000.0);
```
Comment (elrodro83) [inline modules/http-impl/src/main/java/org/mule/services/http/impl/service/server/grizzly/GrizzlyServerManager.java]: can use the method in TimeUnit for this

File: modules/http-impl/src/main/java/org/mule/services/http/impl/service/server/grizzly/GrizzlyServerManager.java [hunk 11]
```diff
@@ -255,11 +249,43 @@ private int retrieveMaximumHeaderSectionSize() {
...
+
+  }
+
+  private class ProxyGrizzlyHttpServer extends GrizzlyHttpServer {
```
Comment (elrodro83) [inline modules/http-impl/src/main/java/org/mule/services/http/impl/service/server/grizzly/GrizzlyServerManager.java]: is this Proxy as in http proxy?

File: modules/http-impl/src/main/java/org/mule/services/http/impl/service/server/grizzly/IdleExecutor.java [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+package org.mule.services.http.impl.service.server.grizzly;
+
+import static java.util.concurrent.Executors.newCachedThreadPool;
+import org.mule.runtime.core.util.concurrent.NamedThreadFactory;
+
+import java.util.concurrent.ExecutorService;
+
+import org.glassfish.grizzly.utils.DelayedExecutor;
+
+public class IdleExecutor {
```
Comment (elrodro83) [inline modules/http-impl/src/main/java/org/mule/services/http/impl/service/server/grizzly/IdleExecutor.java]: javadoc

File: modules/http-impl/src/main/java/org/mule/services/http/impl/service/server/grizzly/IdleExecutor.java [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+
+  private static final String IDLE_TIMEOUT_THREADS_PREFIX_NAME = ".HttpIdleConnectionCloser";
+
+  private ExecutorService idleTimeoutExecutorService;
+  private DelayedExecutor idleTimeoutDelayedExecutor;
+
+  public IdleExecutor(String prefix) {
+    this.idleTimeoutExecutorService = newCachedThreadPool(new NamedThreadFactory(prefix + IDLE_TIMEOUT_THREADS_PREFIX_NAME));
```
Comment (elrodro83) [inline modules/http-impl/src/main/java/org/mule/services/http/impl/service/server/grizzly/IdleExecutor.java]: add TODO MULE-10545

File: modules/http/src/main/java/org/mule/extension/http/internal/listener/HttpResponseContext.java [hunk 14]
```diff
@@ -6,7 +6,9 @@
  */
 package org.mule.extension.http.internal.listener;
 
-import org.mule.runtime.module.http.internal.listener.async.HttpResponseReadyCallback;
+
```
Comment (elrodro83) [inline modules/http/src/main/java/org/mule/extension/http/internal/listener/HttpResponseContext.java]: remove extra lines

File: modules/http/src/main/java/org/mule/extension/http/internal/request/client/HttpExtensionClient.java [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+import org.mule.service.http.api.client.HttpClient;
+import org.mule.service.http.api.client.HttpRequestAuthentication;
+import org.mule.service.http.api.domain.request.HttpRequest;
+import org.mule.service.http.api.domain.response.HttpResponse;
+
+import java.io.IOException;
+import java.util.concurrent.TimeoutException;
+
+/**
+ *
```
Comment (elrodro83) [inline modules/http/src/main/java/org/mule/extension/http/internal/request/client/HttpExtensionClient.java]: javadoc

File: modules/http/src/main/java/org/mule/runtime/module/http/internal/listener/ServerAddressMap.java [hunk 16]
```diff
@@ -35,8 +36,16 @@ public T get(Object key) {
     T value = internalMap.get(key);
     if (value == null) {
       // if there's no entry for the specific address, we need to check if there's one for all interfaces address.
-      value = internalMap.get(new ServerAddress(HttpConstants.ALL_INTERFACES_IP, ((ServerAddress) key).getPort()));
+      value = internalMap.get(new DefaultServerAddress(HttpConstants.ALL_INTERFACES_IP, ((ServerAddress) key).getPort()));
```
Comment (elrodro83) [inline modules/http/src/main/java/org/mule/runtime/module/http/internal/listener/ServerAddressMap.java]: static import

File: modules/http/src/main/java/org/mule/runtime/module/http/internal/listener/grizzly/GrizzlyHttpServer.java [hunk 17]
```diff
@@ -0,0 +1,106 @@
...
+import org.mule.service.http.api.server.ServerAddress;
+
+import java.io.IOException;
+import java.util.concurrent.ExecutorService;
+import java.util.function.Supplier;
+
+import org.glassfish.grizzly.nio.transport.TCPNIOServerConnection;
+import org.glassfish.grizzly.nio.transport.TCPNIOTransport;
+
+public class GrizzlyHttpServer implements HttpServer, MuleContextAware, Supplier<ExecutorService> {
```
Comment (elrodro83) [inline modules/http/src/main/java/org/mule/runtime/module/http/internal/listener/grizzly/GrizzlyHttpServer.java]: javadoc

File: modules/http/src/main/java/org/mule/runtime/module/http/internal/listener/matcher/ListenerRequestMatcher.java [hunk 18]
```diff
@@ -42,6 +44,6 @@ public boolean matches(HttpRequest request) {
 
   @Override
   public String toString() {
-    return "ListenerRequestMatcher{" + "path='" + path + '\'' + ", methodRequestMatcher=" + methodRequestMatcher + '}';
+    return "PathAndMethodRequestMatcher{" + "path='" + path + '\'' + ", methodRequestMatcher=" + methodRequestMatcher + '}';
```
Comment (elrodro83) [inline modules/http/src/main/java/org/mule/runtime/module/http/internal/listener/matcher/ListenerRequestMatcher.java]: why this change? Us the class getSimpleName() method

File: services/http/src/main/java/org/mule/service/http/api/HttpService.java [hunk 19]
```diff
@@ -0,0 +1,19 @@
...
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.service.http.api;
+
+import org.mule.runtime.api.service.Service;
+import org.mule.service.http.api.client.HttpClientFactory;
+import org.mule.service.http.api.server.HttpServerFactory;
+
+public interface HttpService extends Service {
```
Comment (elrodro83) [inline services/http/src/main/java/org/mule/service/http/api/HttpService.java]: javadocs

File: services/http/src/main/java/org/mule/service/http/api/client/HttpClientConfiguration.java [hunk 20]
```diff
@@ -48,14 +40,6 @@ private HttpClientConfiguration(UriParameters uriParameters, HttpAuthentication
     this.ownerName = ownerName;
   }
 
-  public UriParameters getUriParameters() {
```
Comment (elrodro83) [inline services/http/src/main/java/org/mule/service/http/api/client/HttpClientConfiguration.java]: why are these removed?

File: services/http/src/main/java/org/mule/service/http/api/domain/ParameterMap.java [hunk 21]
```diff
@@ -83,7 +83,7 @@ public String get(Object key) {
     return null;
   }
 
-  public List<String> getAll(String key) {
+  public List<String> getAll(Object key) {
```
Comment (elrodro83) [inline services/http/src/main/java/org/mule/service/http/api/domain/ParameterMap.java]: is this needed?

File: services/http/src/main/java/org/mule/service/http/api/domain/ParameterMap.java [hunk 22]
```diff
@@ -83,7 +83,7 @@ public String get(Object key) {
...
     return paramsMap.containsKey(key) ? Collections.unmodifiableList(paramsMap.get(key)) : Collections.<String>emptyList();
```
Comment (elrodro83) [inline services/http/src/main/java/org/mule/service/http/api/domain/ParameterMap.java]: static imports

File: services/http/src/main/java/org/mule/service/http/api/domain/entity/multipart/Part.java [hunk 23]
```diff
@@ -0,0 +1,89 @@
...
+ * LICENSE.txt file.
+ */
+package org.mule.service.http.api.domain.entity.multipart;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Collection;
+
+/**
+ *
```
Comment (elrodro83) [inline services/http/src/main/java/org/mule/service/http/api/domain/entity/multipart/Part.java]: javadoc

File: services/http/src/main/java/org/mule/service/http/api/domain/entity/multipart/Part.java [hunk 24]
```diff
@@ -0,0 +1,89 @@
...
+   * @param name a <code>String</code> specifying the header name
+   *
+   * @return a <code>String</code> containing the value of the requested header, or <code>null</code> if the part does not have a
+   *         header of that name
+   */
+  String getHeader(String name);
+
+  /**
+   * Gets the values of the Part header with the given name.
+   *
```
Comment (elrodro83) [inline services/http/src/main/java/org/mule/service/http/api/domain/entity/multipart/Part.java]: remove empty lines from all javadocs in this interface

File: services/http/src/main/java/org/mule/service/http/api/domain/entity/multipart/Part.java [hunk 25]
```diff
@@ -0,0 +1,89 @@
...
+
+  /**
+   *
+   * Returns the value of the specified mime header as a <code>String</code>. If the Part did not include a header of the
+   * specified name, this method returns <code>null</code>. If there are multiple headers with the same name, this method returns
+   * the first header in the part. The header name is case insensitive. You can use this method with any request header.
+   *
+   * @param name a <code>String</code> specifying the header name
+   *
+   * @return a <code>String</code> containing the value of the requested header, or <code>null</code> if the part does not have a
```
Comment (elrodro83) [inline services/http/src/main/java/org/mule/service/http/api/domain/entity/multipart/Part.java]: Use @code instead of <code>

File: services/http/src/main/java/org/mule/service/http/api/domain/request/ClientConnection.java [hunk 26]
```diff
@@ -0,0 +1,28 @@
...
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.service.http.api.domain.request;
+
+import java.net.InetSocketAddress;
+import java.security.cert.Certificate;
+
+/**
+ *
```
Comment (elrodro83) [inline services/http/src/main/java/org/mule/service/http/api/domain/request/ClientConnection.java]: javadoc

File: services/http/src/main/java/org/mule/service/http/api/domain/request/HttpRequest.java [hunk 27]
```diff
@@ -32,6 +33,8 @@
    */
   String getUri();
 
+  ParameterMap getQueryParams();
```
Comment (elrodro83) [inline services/http/src/main/java/org/mule/service/http/api/domain/request/HttpRequest.java]: javadoc

File: services/http/src/main/java/org/mule/service/http/api/domain/request/HttpRequestContext.java [hunk 28]
```diff
@@ -0,0 +1,29 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.service.http.api.domain.request;
+
+/**
+ *
```
Comment (elrodro83) [inline services/http/src/main/java/org/mule/service/http/api/domain/request/HttpRequestContext.java]: javadco

File: tests/functional/src/main/java/org/mule/functional/junit4/ExtensionFunctionalTestCase.java [hunk 29]
```diff
@@ -80,7 +80,7 @@
    * @param builders the list of {@link ConfigurationBuilder}s that will be used to initialise the {@link #muleContext}
    */
   @Override
-  protected final void addBuilders(List<ConfigurationBuilder> builders) {
+  protected void addBuilders(List<ConfigurationBuilder> builders) {
```
Comment (elrodro83) [inline tests/functional/src/main/java/org/mule/functional/junit4/ExtensionFunctionalTestCase.java]: this could be put back when the circualr dependency issue is removed, right? if so, add a TODO for this

File: tests/http/src/test/java/org/mule/test/module/http/functional/listener/HttpListenerPartsTestCase.java [hunk 30]
```diff
@@ -171,7 +170,7 @@ private String getResponseWithExpectedAttachmentFrom(String path, String require
 
         final Collection<HttpPart> parts = HttpParser.parseMultipartContent(response.getEntity().getContent(), contentType);
         assertThat(parts.size(), is(1));
-        Map<String, Part> partsAsMap = convertPartsToMap(parts);
+        Map<String, org.mule.service.http.api.domain.entity.multipart.Part> partsAsMap = convertPartsToMap(parts);
```
Comment (elrodro83) [inline tests/http/src/test/java/org/mule/test/module/http/functional/listener/HttpListenerPartsTestCase.java]: why fully qualified?

File: modules/http-impl/src/main/java/org/mule/services/http/impl/service/client/GrizzlyHttpClient.java (hunk 6 above)
Comment (afelisatti) [inline modules/http-impl/src/main/java/org/mule/services/http/impl/service/client/GrizzlyHttpClient.java]: Yes, this was extension specific (because host and port could support expression at the config level but we didn't want extra configs/connections being created). I've created the HttpExtensionClient with this and the authentication too.

File: modules/http-impl/src/main/java/org/mule/services/http/impl/service/server/HttpListenerConnectionManager.java (hunk 7 above)
Comment (afelisatti) [inline modules/http-impl/src/main/java/org/mule/services/http/impl/service/server/HttpListenerConnectionManager.java]: No, I need to create one.

File: modules/http-impl/src/main/java/org/mule/services/http/impl/service/server/grizzly/GrizzlyServerManager.java (hunk 9 above)
Comment (afelisatti) [inline modules/http-impl/src/main/java/org/mule/services/http/impl/service/server/grizzly/GrizzlyServerManager.java]: Yes, this was causing a memory leak since having an executor per container caused it's queue to grow without end. I've introduced the IdleExecutor per server address so that they can be destroyed once the server goes down.

File: modules/http-impl/src/main/java/org/mule/services/http/impl/service/server/grizzly/GrizzlyServerManager.java (hunk 11 above)
Comment (afelisatti) [inline modules/http-impl/src/main/java/org/mule/services/http/impl/service/server/grizzly/GrizzlyServerManager.java]: No, maybe it should have a better name. It's just a wrapper to handle the ServerManager specific disposal of per server objects.

File: modules/http-impl/src/main/java/org/mule/services/http/impl/service/server/grizzly/GrizzlyServerManager.java (hunk 10 above)
Comment (afelisatti) [inline modules/http-impl/src/main/java/org/mule/services/http/impl/service/server/grizzly/GrizzlyServerManager.java]: Yes, I think this was a bad rebase.

File: modules/http/src/main/java/org/mule/runtime/module/http/internal/listener/matcher/ListenerRequestMatcher.java (hunk 18 above)
Comment (afelisatti) [inline modules/http/src/main/java/org/mule/runtime/module/http/internal/listener/matcher/ListenerRequestMatcher.java]: I bet it was a bad rename when I was trying some stuff out during the spike.

File: extensions/http/src/main/resources/META-INF/services/org/mule/runtime/core/config/registry-bootstrap.properties [hunk 31]
```diff
@@ -0,0 +1,3 @@
+#_httpMessageProcessorProvider=org.mule.module.http.internal.request.client.HttpConnectorMessageProcessorProvider
```
Comment (elrodro83) [inline extensions/http/src/main/resources/META-INF/services/org/mule/runtime/core/config/registry-bootstrap.properties]: remove or uncomment this line

File: modules/http/src/main/java/org/mule/runtime/module/http/internal/listener/matcher/DefaultMethodRequestMatcher.java [hunk 32]
```diff
@@ -44,13 +44,13 @@ public Object transform(Object input) {
...
    * Determines if there's an intersection between the allowed methods by two request matcher
    *
-   * @param methodRequestMatcher request matcher to test against
+   * @param otherMatcher request matcher to test against
    * @return true at least there's one http method that both request matcher accepts, false otherwise.
    */
-  public boolean intersectsWith(final MethodRequestMatcher methodRequestMatcher) {
-    Preconditions.checkArgument(methodRequestMatcher != null, "methodRequestMatcher cannot be null");
+  public boolean intersectsWith(final MethodRequestMatcher otherMatcher) {
+    Preconditions.checkArgument(otherMatcher != null, "methodRequestMatcher cannot be null");
```
Comment (elrodro83) [inline modules/http/src/main/java/org/mule/runtime/module/http/internal/listener/matcher/DefaultMethodRequestMatcher.java]: update string

File: modules/http/src/main/java/org/mule/runtime/module/http/internal/listener/matcher/AcceptsAllMethodsRequestMatcher.java [hunk 33]
```diff
@@ -27,7 +27,7 @@ public static AcceptsAllMethodsRequestMatcher instance() {
   }
 
   @Override
-  public boolean intersectsWith(MethodRequestMatcher methodRequestMatcher) {
-    return methodRequestMatcher instanceof AcceptsAllMethodsRequestMatcher;
+  public boolean intersectsWith(MethodRequestMatcher otherMatcher) {
+    return otherMatcher instanceof AcceptsAllMethodsRequestMatcher;
```
Comment (elrodro83) [inline modules/http/src/main/java/org/mule/runtime/module/http/internal/listener/matcher/AcceptsAllMethodsRequestMatcher.java]: why not just 'matcher' for the param name?

File: modules/http/src/main/java/org/mule/runtime/module/http/internal/listener/matcher/DefaultMethodRequestMatcher.java [hunk 34]
```diff
@@ -44,13 +44,13 @@ public Object transform(Object input) {
   /**
    * Determines if there's an intersection between the allowed methods by two request matcher
    *
-   * @param methodRequestMatcher request matcher to test against
+   * @param otherMatcher request matcher to test against
    * @return true at least there's one http method that both request matcher accepts, false otherwise.
    */
-  public boolean intersectsWith(final MethodRequestMatcher methodRequestMatcher) {
-    Preconditions.checkArgument(methodRequestMatcher != null, "methodRequestMatcher cannot be null");
+  public boolean intersectsWith(final MethodRequestMatcher otherMatcher) {
```
Comment (elrodro83) [inline modules/http/src/main/java/org/mule/runtime/module/http/internal/listener/matcher/DefaultMethodRequestMatcher.java]: why not just 'matcher' for the param name?

File: services/http/src/main/java/org/mule/service/http/api/client/HttpClient.java [hunk 35]
```diff
@@ -19,12 +19,18 @@
  */
 public interface HttpClient {
 
+  /**
+   * Fully configures the client, leaving it ready to use.
+   */
   void start();
 
+  /**
+   * Disables the client.
```
Comment (elrodro83) [inline services/http/src/main/java/org/mule/service/http/api/client/HttpClient.java]: add in this javadoc whether it should dispose resources, stop threads, etc

File: services/http/src/main/java/org/mule/service/http/api/domain/entity/multipart/Part.java [hunk 36]
```diff
@@ -50,35 +52,29 @@
    * specified name, this method returns <code>null</code>. If there are multiple headers with the same name, this method returns
```
Comment (elrodro83) [inline services/http/src/main/java/org/mule/service/http/api/domain/entity/multipart/Part.java]: `@code`. review other occurrences in this interface

File: services/http/src/main/java/org/mule/service/http/api/domain/entity/multipart/Part.java [hunk 37]
```diff
@@ -50,35 +52,29 @@
...
    * the first header in the part. The header name is case insensitive. You can use this method with any request header.
    *
-   * @param name a <code>String</code> specifying the header name
-   *
-   * @return a <code>String</code> containing the value of the requested header, or <code>null</code> if the part does not have a
+   * @param name a {@code String} specifying the header name
```
Comment (elrodro83) [inline services/http/src/main/java/org/mule/service/http/api/domain/entity/multipart/Part.java]: for java classes like String use `@link`

File: services/http/src/main/java/org/mule/service/http/api/domain/response/HttpResponse.java [hunk 38]
```diff
@@ -8,12 +8,24 @@
...
+ */
 public interface HttpResponse extends HttpMessage {
 
+  /**
+   * @return the HTTP status line code
+   */
   int getStatusCode();
 
+  //TODO - MULE-10760: Remove setters once builders are available
   void setStatusCode(int statusCode);
```
Comment (elrodro83) [inline services/http/src/main/java/org/mule/service/http/api/domain/response/HttpResponse.java]: deprecate setters

Comment (afelisatti) [inline services/http/src/main/java/org/mule/service/http/api/domain/response/HttpResponse.java]: They won't live longer than a few days, I don't think it makes sense.

File: extensions/http/src/main/resources/META-INF/services/org/mule/runtime/core/config/registry-bootstrap.properties [hunk 39]
```diff
@@ -1,4 +1,3 @@
-_httpExtListenerConnectionManager=org.mule.extension.http.internal.listener.server.HttpListenerConnectionManager,applyToArtifactType=app/domain
 #_httpMessageProcessorProvider=org.mule.module.http.internal.request.client.HttpConnectorMessageProcessorProvider
```
Comment (elrodro83) [inline extensions/http/src/main/resources/META-INF/services/org/mule/runtime/core/config/registry-bootstrap.properties]: remove or uncomment

File: modules/http-api/src/main/java/org/mule/service/http/api/domain/entity/InputStreamHttpEntity.java [hunk 40]
```diff
@@ -4,10 +4,15 @@
  * license, a copy of which has been included with this distribution in the
  * LICENSE.txt file.
  */
-package org.mule.runtime.module.http.internal.domain;
+package org.mule.service.http.api.domain.entity;
 
 import java.io.InputStream;
 
+/**
+ * Representation of an stream HTTP body.
```
Comment (elrodro83) [inline modules/http-api/src/main/java/org/mule/service/http/api/domain/entity/InputStreamHttpEntity.java]: an -> a

File: modules/http-api/src/main/java/org/mule/service/http/api/domain/entity/multipart/Part.java [hunk 41]
```diff
@@ -0,0 +1,85 @@
...
+
+/**
+ * Represent a part of a multipart body.
+ *
+ * @since 4.0
+ */
+public interface Part {
+
+  /**
+   * Gets the content of this part as an <tt>InputStream</tt>
```
Comment (elrodro83) [inline modules/http-api/src/main/java/org/mule/service/http/api/domain/entity/multipart/Part.java]: tt -> `@link`

File: modules/http-api/src/main/java/org/mule/service/http/api/domain/entity/multipart/Part.java [hunk 42]
```diff
@@ -0,0 +1,85 @@
...
+   * @param name a {@code String} specifying the header name
+   * @return a {@code String} containing the value of the requested header, or {@code null} if the part does not have a
+   *         header of that name
+   */
+  String getHeader(String name);
+
+  /**
+   * Gets the values of the Part header with the given name.
+   * <p>
+   * Any changes to the returned <code>Collection</code> must not affect this <code>Part</code>.
```
Comment (elrodro83) [inline modules/http-api/src/main/java/org/mule/service/http/api/domain/entity/multipart/Part.java]: code -> `@link`

File: modules/http-api/src/main/java/org/mule/service/http/api/domain/response/HttpResponse.java [hunk 43]
```diff
@@ -4,16 +4,28 @@
...
+ * @since 4.0
+ */
 public interface HttpResponse extends HttpMessage {
 
+  /**
+   * @return the HTTP status line code
+   */
   int getStatusCode();
 
+  //TODO - MULE-10760: Remove setters once builders are available
```
Comment (elrodro83) [inline modules/http-api/src/main/java/org/mule/service/http/api/domain/response/HttpResponse.java]: deprecate setters

File: modules/http-api/src/main/java/org/mule/service/http/api/server/ServerAddress.java [hunk 44]
```diff
@@ -0,0 +1,31 @@
...
+
+/**
+ * Representation of a server's address.
+ *
+ * @since 4.0
+ */
+public interface ServerAddress {
+
+  /**
+   * @return the port this server is bind to.
```
Comment (elrodro83) [inline modules/http-api/src/main/java/org/mule/service/http/api/server/ServerAddress.java]: bound

File: pom.xml [hunk 45]
```diff
@@ -99,11 +99,11 @@
         <module>buildtools</module>
         <module>core</module>
         <module>core-tests</module>
-        <module>services</module>
         <module>modules</module>
         <module>tests</module>
         <module>extensions</module>
         <module>transports</module>
+        <module>services</module>
```
Comment (elrodro83) [inline pom.xml]: is this reorder needed?

File: services/http/src/main/java/org/mule/services/http/impl/service/server/grizzly/GrizzlyServerManager.java [hunk 46]
```diff
@@ -255,11 +250,43 @@ private int retrieveMaximumHeaderSectionSize() {
...
+
+    @Override
+    public synchronized void start() throws IOException {
+      idleExecutorPerServerAddressMap.get(this.getServerAddress()).start();
+      super.start();
+    }
+
+    @Override
+    public synchronized void dispose() {
+      super.dispose();
```
Comment (elrodro83) [inline services/http/src/main/java/org/mule/services/http/impl/service/server/grizzly/GrizzlyServerManager.java]: should this shutdown the idle executor too?

File: pom.xml (hunk 45 above)
Comment (afelisatti) [inline pom.xml]: I don't think so, I'll check, thanks.

File: services/http/src/main/java/org/mule/services/http/impl/service/server/grizzly/GrizzlyServerManager.java (hunk 46 above)
Comment (afelisatti) [inline services/http/src/main/java/org/mule/services/http/impl/service/server/grizzly/GrizzlyServerManager.java]: it's done below in two parts, first doing the dispose and then removing the reference.

File: extensions/http/src/main/resources/META-INF/services/org/mule/runtime/core/config/registry-bootstrap.properties (hunk 39 above)
Comment (afelisatti) [inline extensions/http/src/main/resources/META-INF/services/org/mule/runtime/core/config/registry-bootstrap.properties]: Though this was used by the mule client, I'll remove it and in any case we can reintroduce it if it's necessary for the revamped mule client.

File: modules/http-api/src/main/java/org/mule/service/http/api/domain/entity/multipart/Part.java [hunk 47]
```diff
@@ -35,21 +35,21 @@
...
-   * @return a <code>long</code> specifying the size of this part, in bytes.
+   * @return a {@code long} specifying the size of this part, in bytes.
    */
   long getSize();
 
   /**
    *
-   * Returns the value of the specified mime header as a <code>String</code>. If the Part did not include a header of the
-   * specified name, this method returns <code>null</code>. If there are multiple headers with the same name, this method returns
+   * Returns the value of the specified mime header as a {@code String}. If the Part did not include a header of the
```
Comment (elrodro83) [inline modules/http-api/src/main/java/org/mule/service/http/api/domain/entity/multipart/Part.java]: `@link`

File: modules/http-api/src/main/java/org/mule/service/http/api/domain/entity/multipart/Part.java [hunk 48]
```diff
@@ -61,24 +61,24 @@
   /**
    * Gets the values of the Part header with the given name.
    * <p>
-   * Any changes to the returned <code>Collection</code> must not affect this <code>Part</code>.
+   * Any changes to the returned {@code Collection} must not affect this {@code Part}.
```
Comment (elrodro83) [inline modules/http-api/src/main/java/org/mule/service/http/api/domain/entity/multipart/Part.java]: `@link`

File: modules/http-api/src/main/java/org/mule/service/http/api/domain/entity/multipart/Part.java [hunk 49]
```diff
@@ -61,24 +61,24 @@
...
    * <p>
    * Part header names are case insensitive.
    *
    * @param name the header name whose values to return
-   * @return a (possibly empty) <code>Collection</code> of the values of the header with the given name
+   * @return a (possibly empty) {@code Collection} of the values of the header with the given name
```
Comment (elrodro83) [inline modules/http-api/src/main/java/org/mule/service/http/api/domain/entity/multipart/Part.java]: `@link`

File: modules/http-api/src/main/java/org/mule/service/http/api/domain/entity/multipart/Part.java [hunk 50]
```diff
@@ -61,24 +61,24 @@
...
 
   /**
    * Gets the header names of this Part.
    * <p>
    * Some servlet containers do not allow servlets to access headers using this method, in which case this method returns
-   * <code>null</code>
+   * {@code null}
    * <p>
-   * Any changes to the returned <code>Collection</code> must not affect this <code>Part</code>.
+   * Any changes to the returned {@code Collection} must not affect this {@code Part}.
```
Comment (elrodro83) [inline modules/http-api/src/main/java/org/mule/service/http/api/domain/entity/multipart/Part.java]: `@link`

File: modules/http-api/src/main/java/org/mule/service/http/api/domain/entity/multipart/Part.java [hunk 51]
```diff
@@ -61,24 +61,24 @@
...
    *
-   * @return a (possibly empty) <code>Collection</code> of the header names of this Part
+   * @return a (possibly empty) {@code Collection} of the header names of this Part
```
Comment (elrodro83) [inline modules/http-api/src/main/java/org/mule/service/http/api/domain/entity/multipart/Part.java]: `@link`
